prov/efa/docs/efa_rdm_protocol_v4.md

```diff
@@ -0,0 +1,1318 @@
+
+However, there is no guarantee in the base protocol that the packet header length of EAGER_MSGRTM will not
+change.
+
+In fact, because of the existance of the handshake sub-protocol, the packet header length of an EAGER_MSGRTM
+will definitely change. Recall handshake sub-protocol's workflow is:
+
+Before receiving handshake packet, an endpoint will always include the optional raw address header in REQ packets.
+
+After receiving handshake packet, an endpoint will stop including the optional raw address header in REQ packets.
+
+The extra feature "keep packet header length stable" (stable header length) is designed to solve this problem.
+
+When an endpoint toggle on this extra request, its peer will try to satisfy it by keep the header length stable.
+Exactly how to achieve that is up to the implementation to decide, the easiest way to do that is to keep including
+raw address header in the EAGER_MSGRTM even after receiving the handshake packet.
+
+Note, because this is an extra request, an endpoint cannot assume its peer will comply with the request. Therefore,
+the receiving endpoint must be able to handle the situation that a received packet does not have the expected header length.
+
+In that case, implementation will have two choices:
+
+1. write a truncated error completion entry
+2. move the application data to right place.
+
+Note this extra request was initially introduced as an extra feature named "zero copy receive", but it is later realized
+that is not an feature because the peer does not do anything different. Rather, it is an expectation the receiving
+endpoint has for the sender. Therefore, it was re-interpreted as an extra request named "stale header length".
+This re-interpretation does not change the implementation, thus does not cause backward incompatibility.
+
+### 4.4 have connection ID in packet header (connid header)
+
+The "have connection ID in packet header" extra request was introduced with libfabric 1.14.0 release, and was
+assigned the ID 3.
+
+This extra feature is designed to solve the "QP collision" problem, which is commonly experienced in
+client-server type of application.
+
+The "QP collision" problem arise from fact that the EFA device uses the gid + qpn as the unique
+identifier of a peer. Recall that raw address of EFA endpoint is consisted of 3 parts:
+gid + qpn + connid. EFA device only recognizes gid and qpn. The connid was generated by the endpoint
+itself during its initialization.
+
+Because of that, it is possible for an endpoint to receive packets from a destroyed
+peer. The endpoint does not know how to handle such a packet, and will crash or cause data corruption.
+
+The problem can be demonstrated using the following workflow:
+
+    1. We have two instances A and B.
+    2. On instance A we create endpoint 1.
+    3. On instance B we create endpoint 2.
+    4. Endpoint 1 and 2 start communication with each other.
+    5. There is an accident on instance B, endpoint 2 was destroyed.
```

Comment:
"Instance B crashes, endpoint 2 is destroyed"

Reply:
I added `application crashed, hence endpoint 2 was destroyed`.  The instance itself might be OK in this case.